### ISSUE TITLE
ThirdStats counts are always incorrect

### ISSUE THREAD

**Issue by desrod** [devmount/third-stats#413]
I was hoping ThirdStats would show the details of how many mails per day/week/month/year, but the counts are always wrong. 

Currently, ThirdStats claims I'm getting 9.2 emails per-day in the last 1 month. 

![2023-06-02_15-01](https://github.com/devmount/third-stats/assets/5942125/a1aff4cf-f294-4a38-88ea-694f197add74)

But looking at Thunderbird directly, the counts are _significantly_ higher than that: 

![2023-06-02_15-03](https://github.com/devmount/third-stats/assets/5942125/d37f13ad-bb38-493f-b35b-ab63bef0ef28)

I've dumped the cache, re-built, tried a new profile and so on, all with the same results. 

How can I get ThirdStats to show accurate statistics?

**Comment by devmount** [comment]
Hi David,

thank you for this report. What do you mean by _looking at Thunderbird directly_? How do you create the view in your second screenshot. Also the numbers in your second screenshot seem odd. How can there be more emails in the last 7 days than in the last 14 days (which I assume must include the last 7 days).

ThirdStats only uses Thunderbirds official WebExtension APIs - so the counts reflect all messages retrieved over those APIs.

**Comment by desrod** [comment]
@devmount When I view the folder, threaded, Grouped by Sort, and sort by date, then collapse those folders, I get those counts. It's all built-in.

**Comment by devmount** [comment]
First of all: I cannot reproduce your problem. I created the view like you did:

![image](https://github.com/devmount/third-stats/assets/5441654/c1ae928a-ccd2-4adf-ac28-1084e11d96a8)

And compared with ThirdStats for this folder:

![image](https://github.com/devmount/third-stats/assets/5441654/3c25c405-6f78-4262-98a0-8c3248268b85)

(sorry it's German) The numbers match on my end.

Things you can do:

1. Check that you have the correct account selected on the stats page
2. Check that you have the correct folder selected on the stats page
3. Check, if all messages of that folder are actually downloaded by Thunderbird (not only the message headers - I used to experience some number differences, when Thunderbird only had message headers stored due to performance reasons, which weren't delivered over webextension APIs)

**Comment by desrod** [comment]
I've confirmed this with 2 cleanly created profiles on 2 separate machines, using the same version of Thunderbird on both (macOS and my daily Linux machine), and the results were the same. 

The counts were 10x to 100x below the actual count of mail in each of the selected folders. All mail was fully sync'd to the local machine, and it's the same whether I'm "offline" or online, when re-generating the statistics. 

I have 8 different mail accounts that I use, and each of them has the same results, none of the counts were accurate across any of them. It's definitely consistent across OS/platform and accounts. 

I think I've noticed at least one part of why this is thrown off. For my work account, I receive quite a bit less email on the weekends (~200 emails a day on weekdays, maybe 10-50 on the weekend days), and this throws the statistics off by quite a bit as I include prior years. 

Having a "Days" count next to the Years | Months | Weeks would make this a lot easier to understand as well. I'll see if I can't PR that in. Excluding weekends from the parsing would also be useful, on the main stats screen (a simple checkbox: "Exclude weekend days" for example). 

I think this might be why the stats are progressively off the further back the date constraint goes. 

Thanks for testing for me!

**Comment by devmount** [comment]
This is interesting. Though I can't think of any way how this could be connected to weekends. Just to make sure: You don't have any warnings when processing like this?

![thirdstats_errorhandling](https://user-images.githubusercontent.com/5441654/227018752-657bad1d-330b-40e0-ac98-a59bff7088b2.png)

Which email provider(s) do you use? Several users had problems with virtual folders from some providers (e.g. gmail) in the past.

**Comment by devmount** [comment]
Feel free to reopen this, if you have any news on this.

**Comment by desrod** [comment]
Thanks for your patience. The issue still persists with any account I have tested with (7 separate Gmail accounts). 

Just looking at yesterday's incoming mail to my my most heavily-used account, there are 236 emails in my Inbox received on that date, where my email address is directly on the 'To:' line (not Bcc nor Cc, and not from a mailing list). 

Searching exclusively in my Inbox, ThirdStats says I'm receiving 59.7 emails per-day (after clearing and rebuilding the cache). There are over 1,500 emails in my Inbox (intentionally not cleaning it out, in prep for a talk about email). 

The counts just don't seem to be consistent. 

Is there some debug mode I can run in and figure out precisely what math it's using to calculate these, and where it's going off trail and miscounting them?

**Comment by devmount** [comment]
There is no built in debug mode for ThirdStats, but actually that's a very good idea. We could add a debug mode switch in the settings which logs verbose messages to the console if enabled. I created #447 for this.

**Comment by devmount** [comment]
I implemented the debug mode now and released ThirdStats 1.12.0 - you can wait until its approved on ATN or download the XPI file from the release and install it manually to test it out. Let me know if that was helpful in any way for your issue.

**Comment by ajoe** [comment]
I have the same "stats are incorrect issue" only on Windows 11. The same Thunderbird setup on macOS (M2) works fine. Thunderbird 137 on both. 
How can I enable the debug mode?

**Comment by devmount** [comment]
@ajoe That's interesting, thanks for reporting those numbers! Unfortunately until now I wasn't able to reproduce that on my machines (Windows 10, Ubuntu-based Linux).
You can enable the debug mode in the ThirdStats Options (sorry for the German screenshots):

1. Open the options page (cog symbol upper right corner in the popup or on the stats page
    ![Image](https://github.com/user-attachments/assets/d15e1a7f-13aa-4211-ab38-10a0e8bd7434)
2. Scroll down and enable the debug mode toggle
    ![Image](https://github.com/user-attachments/assets/f1384402-8bf6-4f04-9144-583b1003e501)
3. Open the developer console (add-ons > gear > debug add-ons > scroll to ThirdStats > Inspect) and click the refresh button on the stats page

Now there should be verbose output how many emails were detected in each folder.

**Comment by ajoe** [comment]
Thanks!
Can you also add the total number for each folder, so not cumulating. Because that is what is shown in Thunderbird as a per folder count.

**Comment by ajoe** [comment]
I can see a differenz. 
ThirdStats shows total count (A, B are foldernames): 
A / A / A = 206
A / B / A = 364 This is wrong. There are not 158 mails. Only 58 are counted in Thunderbrid. But they are shown Threaded. If Threads are disabeled, Thunderbird only counts 36 topic.

**Comment by ajoe** [comment]
We found the problem. Thunderbird "Virtual Folder" or "Search-Folder" are included in the statistics of ThirdStats. If you change the search-folder in a way that the find zero / 0 mail, then the count between Thunderbird and ThirdStats is the same. 

What would help, if the debug output would also display the count for every folder. The cumulation counting is hard to compare with the numbers in Thunderbird.

**Comment by devmount** [comment]
@ajoe Thank you so much for investigating this! I'll take care of the next steps to exclude search folders from stats processing.

I will change the debug output like you suggested to only show folder counts instead of accumulated numbers.

**Comment by ajoe** [comment]
Yes. My counting problem with the virtual folder is solved! Also, the debug output is good now.

**Comment by devmount** [comment]
Perfect. Thanks for your help on nailing this issue down.
